### PR TITLE
feat: add Gemma4 E2B multimodal (text/image/audio) on ExecuTorch with XNNPACK FP32 + 8da4w

### PR DESCRIPTION
**Pull request by seyeong-han**
**Working on CUDA backend**

## Summary

Adds a complete ExecuTorch implementation of `google/gemma-4-E2B-it` (2B parameter multimodal LLM) supporting text, image+text, and audio+text generation from a single `.pte` artifact via the standard `MultimodalRunner`. Generates clean, coherent output across all three modalities; layer-by-layer parity with HuggingFace verified; FP32 and 8-bit-dynamic-activation / 4-bit-weight (8da4w) variants both pass 9/10 end-to-end tests.

## What's new

- **`examples/models/gemma4/`** — full model implementation
  - `__init__.py`, `convert_weights.py` — `Gemma4Model` (`Llama2Model` subclass) + HF→ET weight conversion
  - `config/e2b_config.json` — full E2B config (35 layers, dual RoPE, YOCO, PLI, mixed head dims)
  - `encoders.py` — `VisionEncoderExport`, `AudioEncoderExport` wrappers (patchification + position grid in graph)
  - `audio_preprocessor.py` — `Gemma4AudioPreprocessor` (PCM → log-mel, dynamic length)
  - `export_gemma4_multimodal.py` — single-script export of the 5-method `.pte` (`token_embedding`, `text_decoder`, `vision_encoder`, `audio_preprocessor`, `audio_encoder`); `--qmode 8da4w --group-size 32` flags for quantization
  - `main.cpp`, `CMakeLists.txt`, `CMakePresets.json` — C++ runner with `Gemma4DecoderRunner` (passes `pli_token_ids` as 3rd input to `text_decoder`); CPU+XNNPACK and CUDA build presets; `make gemma4-cpu` / `make gemma4-cuda` Makefile targets
  - `chat_template.jinja`, `render_chat.py` — official vLLM Gemma4 chat template
  - `tests/test_parity.py` — 6 layer-by-layer parity tests vs HF
  - `tests/test_kvcache_decode.py`, `tests/test_textdec_wrapper.py` — KV-cache + multimodal-wrapper bit-exact regression tests
  - `test_multimodal.sh` — end-to-end test harness for all 3 modalities
  - `README.md`, `PROGRESS.md`, `TEST_RESULTS.md`, `HANDOVER.md`, `CUDA_HANDOVER.md` — user-facing docs and continuity notes

- **Extensions to shared `examples/models/llama/` infrastructure** — Gemma4-specific features added behind feature-flag args, no behavior change for existing models:
  - `model_args.py` — new args: `global_rope_theta`, `global_partial_rotary_factor`, `global_rope_type`, `hidden_size_per_layer_input`, `global_head_dim`, `attention_multiplier`, `use_v_norm`, `use_layer_scalar`, `act_fn`
  - `rope.py` — dual-buffer RoPE (sliding θ=10k partial=1.0 vs full θ=1e6 partial=0.25 `proportional`); per-layer-type freqs lookup
  - `attention.py` — `attention_multiplier` (Gemma4 uses 1.0, no implicit `1/sqrt(head_dim)`); `global_head_dim` per layer; `v_norm`; YOCO type-aware shared K/V routing
  - `feed_forward.py` — `act_fn` arg (Gemma4 uses `gelu_approx`)
  - `llama_transformer.py` — PLI plumbing (per-layer input gating, 256-dim bottleneck per layer, sourced from a separate embedding table); per-layer-type RoPE; YOCO donor map; `post_attention_norm` + `post_ffn_norm`; `layer_scalar`; embedding scale `sqrt(hidden_size)`; final logit softcap `c·tanh(logits/c)`. **Reads `pli_token_ids` from `attn_options` so the multimodal `h=embeds` path computes PLI correctly** (this fix was the root cause of v6–v10 multimodal degeneration).
  - `source_transformation/sdpa.py` — forward `scale` parameter to `torch.ops.llama.custom_sdpa`
  - `source_transformation/custom_kv_cache.py` — minor compatibility tweak
  - `export_llama_lib.py` — register `gemma4` in `EXECUTORCH_DEFINED_MODELS`

- **`extension/llm/runner/multimodal_prefiller.cpp`** — builds `pli_ids` per input branch (text → real token IDs; image/audio → `pad_token_id=0` per HF `Gemma4Model.forward` line 2215) and conditionally passes them as the 3rd input to `text_decoder` when the `.pte` has the 3-input signature. Falls back transparently to the 2-input ABI for older `.pte`s.

- **`extension/llm/export/config/llm_config.py`** — adds `ModelType.gemma4`.

### Layer-by-layer parity vs HuggingFace (FP32)

`examples/models/gemma4/tests/test_parity.py` — all six tests pass:

| Test | max_abs_diff | cos_sim |
|------|--------------|---------|
| token_embedding | 0.00e+00 | 1.000001 |
| rmsnorm | 0.00e+00 | 1.000000 |
| pli_inputs | 0.00e+00 | 1.000003 |
| rope (sliding cos+sin) | 0.00e+00 | 1.000000 |
| rope (full cos+sin) | 0.00e+00 | 1.000000 |
| decoder_layer_0 | 0.00e+00 | 1.000001 |
| full_forward (logits) | 5.72e-06 | 0.999999 |

Top-1 next-token matches HF `Gemma4ForConditionalGeneration` for the canonical prompt.

### End-to-end multimodal (CPU + XNNPACK)

| Modality | FP32 prefill / decode | 8da4w prefill / decode |
|----------|-----------------------|------------------------|
| Text     | 89–100 / 12.6–13.1 tok/s | 108–122 / 15.9–16.5 tok/s |
| Image    | 92–96 / 12.0–13.0 tok/s | 100–102 / 14.1–14.6 tok/s |
| Audio    | 254 / 11.3 tok/s | 311 / 13.7 tok/s |

`.pte` size: FP32 21 GB, 8da4w 13 GB (35% smaller). 9/10 tests pass for both variants. Sample outputs:

```
Text  "What is 12 multiplied by 8?"
      → "12 multiplied by 8 is **96**.<turn|>"

Image "Describe this image in detail." (strawberry photo)
      → "This is a close-up photograph featuring a single, vibrant red strawberry...
         **Subject (Strawberry):** ...plump, ripe strawberry... **Background:** ...
         heavily blurred (bokeh effect)..."

Audio "Transcribe this audio." (20s Obama farewell)
      → "This week I traveled to Chicago to deliver my final farewell address to
         the nation, following in the tradition of presidents before me..."
```

The 1 failure is a 2 s audio clip — encoder design requires ≥ 5 s of mel context, not a model bug. Full results in `examples/models/gemma4/TEST_RESULTS.md`.

## Review order (large PR)

1. **`examples/models/gemma4/`** — the new code; review on its own
2. **`examples/models/llama/`** Gemma4-specific extensions — feature-flagged, but please verify they don't regress llama/llama3/qwen
3. **`extension/llm/runner/multimodal_prefiller.cpp`** — PLI 3rd input is opt-in via `num_inputs >= 3` check
4. **`extension/llm/export/config/llm_config.py`** — `ModelType.gemma4` enum addition only

## Test plan

- [ ] `pytest examples/models/gemma4/tests/test_parity.py -v` — six layer-by-layer parity tests pass
- [ ] `python examples/models/gemma4/tests/test_textdec_wrapper.py` — wrapper-vs-tokens path bit-exact (`max_diff = 0.0`)
- [ ] FP32 export: `python examples/models/gemma4/export_gemma4_multimodal.py --hf-model ... --et-checkpoint ... --output gemma4_mm.pte --backend xnnpack --max-seq-len 1024 --audio-frames 1976`
- [ ] 8da4w export: same command + `--qmode 8da4w --group-size 32`
- [ ] `make gemma4-cpu`
- [ ] `./examples/models/gemma4/test_multimodal.sh` — 9/10 tests pass on the resulting `.pte`
- [ ] No regression on existing llama/llama3 export flows (smoke test the canonical "What is the capital of France?" path)

## Notes

- Authored with Claude Code.
- The branch also contains preceding commits (`a42f1fbd05`, `5a7b515f8c`) that add a Voxtral TTS prototype. **If reviewers prefer**, those can be split into a separate PR by rebasing this branch onto `main` minus those two commits — happy to do so. Otherwise they ride along since they touch disjoint files (`examples/models/voxtral_tts/`).
- CUDA backend wiring is not in this PR; `examples/models/gemma4/CUDA_HANDOVER.md` documents the path forward (~20-line addition mirroring `examples/models/dinov2/export_dinov2.py`).

### CODE REVIEW

**Comment by digantdesai** [comment]
@claude Review this PR
